server/socketManager.js

```diff
@@ -0,0 +1,333 @@
+   * @returns {null}
+   */
+  handleSocket(socket) {
+    socket.hook('auth', (...args) => {
+      if (socket.channel('auth')) throw new Error('Already authed.');
+
+      if (!this.options.noAuth) {
+        const valid = new Validator(args);
+        const key = valid.str(0, '');
+        if (key.length > 255) throw new Error('Invalid API key.');
+
+        const data = Buffer.from(key, 'utf8');
+        const hash = sha256.digest(data);
+
+        if (!ccmp(hash, this.options.apiHash))
+          throw new Error('Invalid API key.');
+      }
+
+      socket.join('auth');
+      this.logger.info('Successful auth from %s.', socket.host);
+      this.handleAuth(socket);
+      return null;
+    });
+  }
+
+  /*
+   * Utility to verify that the pathname
+   * matches with the id of a client on the class
+   * @private
+   * @param {string} pathname - pathname to validate
+   * @returns {string} id
+   */
+  getIdFromPath(pathname) {
+    assert(
+      pathname[0] === '/' && pathname[pathname.length - 1] === '/',
+      'Invalid pathname'
+    );
+    const id = pathname.slice(1, pathname.length - 1);
+    if (!this.clients.has(id))
+      this.logger.warning(
+        `getIdFromPath: id ${id} from pathname ${pathname} does not exist`
+      );
+
+    return id;
+  }
+
+  /*
+   * Handle new auth'd websocket.
+   * @private
+   * @param {WebSocket} socket
+   */
+  handleAuth(socket) {
+    this.logger.info(`Preparing socket from ${socket.url}`);
+    const parseEvent = event => {
+      assert(typeof event === 'string', 'Must pass a string to getClient');
+      // defaulting to node client if there is no client prefix
+      let parsedEvent = event;
+      let id = 'node';
+
+      // check if first full word is a valid type
+      const split = event.indexOf(' ');
+      const prefix = event.slice(0, split);
+      if (this.types.indexOf(prefix) !== -1) {
+        id = prefix;
+        parsedEvent = event.slice(split + 1);
+      }
+      return [id, parsedEvent];
+    };
+
+    // pathname will be used similar to socket.io's namespaces
+    // they should correspond to the client id that will be used
+    // to make the calls
+    const { pathname } = new URL(socket.ws.url);
+    const id = this.getIdFromPath(pathname);
+    // broadcasts only send messages to the  node
+    // but originating socket does not expect a response
+    /**
+      // Example broadcast from client:
+      socket.fire('broadcast', 'set filter', '00000000000000000000');
+      // which will result in the following fire to bcoin server
+      nodeClient.socket.fire('set filter', '00000000000000000000');
+    **/
+    socket.bind('broadcast', (_event, ...args) => {
+      assert(
+        this.clients.has(id),
+        `No client ${id} for request from ${socket.url}`
+      );
+      const [clientId, event] = parseEvent(_event);
+      const client = this.clients.get(id)[clientId];
+      this.logger.info(`broadcast "${event}"" to ${id}'s ${clientId} client`);
```

Comment:
`this.logger.info(`broadcast "${event}"" to ${id}'s ${clientId} client`);`
vs.
`this.logger.info(`broadcast "%s" to %s's %s client`, event, id, clientId);`

We should look at the logger source code to determine if using the
`%s` syntax will only format the string if the log level has been set to
the same or lower level, if so, it would be a good habit to get into using
that syntax as it would be slightly more performant, even though we aren't
hitting performance bottlenecks, I think it would just be a good general
practice of using logging

Reply:
Ah good call! I'll switch them over 👍

Reply:
I generally prefer the template literal way as I find it way more readable but I think you're right in terms of performance.